metaflow/plugins/tag_cli.py

```diff
@@ -0,0 +1,531 @@
+            )
+            msg += "\n".join(["\t*%s*" % tag for tag in user_identifier_tags])
+
+        raise CommandException(msg)
+
+    return run
+
+
+@click.group()
+def cli():
+    pass
+
+
+@cli.group(help="Commands related to Metaflow tagging.")
+def tag():
+    pass
+
+
+@tag.command("add", help="Add tags to a run.")
+@click.option(
+    "--run-id",
+    required=False,  # set False here so we can throw a better error message
+    default=None,
+    type=str,
+    help="Run ID of the specific run to tag. [required]",
+)
+@click.option(
+    "--namespace",
+    "user_namespace",
+    required=False,
+    default=None,
+    type=str,
+    help="Change namespace from the default (your username) to the one specified.",
+)
+@click.argument("tags", required=True, type=str, nargs=-1)
+@click.pass_obj
+def add(obj, run_id, user_namespace, tags):
+    user_namespace = resolve_identity() if user_namespace is None else user_namespace
+    run = _get_client_run_obj(obj, run_id, user_namespace)
+
+    run.add_tags(tags)
+
+    obj.echo("Operation successful. New tags:", err=False)
+    _print_tags_for_one_run(obj, run)
+
+
+@tag.command("remove", help="Remove tags from a run.")
+@click.option(
+    "--run-id",
+    required=False,  # set False here so we can throw a better error message
+    default=None,
+    type=str,
+    help="Run ID of the specific run to tag. [required]",
+)
+@click.option(
+    "--namespace",
+    "user_namespace",
+    required=False,
+    default=None,
+    type=str,
+    help="Change namespace from the default (your username) to the one specified.",
+)
+@click.argument("tags", required=True, type=str, nargs=-1)
+@click.pass_obj
+def remove(obj, run_id, user_namespace, tags):
+    user_namespace = resolve_identity() if user_namespace is None else user_namespace
+    run = _get_client_run_obj(obj, run_id, user_namespace)
+
+    run.remove_tags(tags)
+
+    obj.echo("Operation successful. New tags:")
+    _print_tags_for_one_run(obj, run)
+
+
+@tag.command("replace", help="Replace one tag of a run.")
+@click.option(
+    "--run-id",
+    required=False,  # set False here, so we can throw a better error message
+    default=None,
+    type=str,
+    help="Run ID of the specific run to tag. [required]",
+)
+@click.option(
+    "--namespace",
+    "user_namespace",
+    required=False,
+    default=None,
+    type=str,
+    help="Change namespace from the default (your username) to the one specified.",
+)
+@click.argument("tags", required=True, type=str, nargs=2)
+@click.pass_obj
+def replace(obj, run_id, user_namespace, tags):
+    user_namespace = resolve_identity() if user_namespace is None else user_namespace
+    run = _get_client_run_obj(obj, run_id, user_namespace)
+
+    run.replace_tag(tags[0], tags[1])
+
+    obj.echo("Operation successful. New tags:")
+    _print_tags_for_one_run(obj, run)
+
+
+# TODO(jackie) confirm with team on whether this can supercede "replace" command above
```

Comment:
@romain-intel  @savingoyal  Most of this CLI was lifted from this [old PR](https://github.com/Netflix/metaflow/pull/547) (as suggested by @romain-intel  to be the latest local implementation in use).  Mostly minor changes (vendor click, some variable renames, style stuff).

In this draft PR, there are two subcommands related to tag replacement:

* `tag replace tag_to_add tag_to_remove` - comes from this [PR](https://github.com/Netflix/metaflow/pull/547)
* `tag replace-many [--add tag_to_add]+ [--remove tag_to_remove]` - a suggested substitute for the `tag replace` above.

For OSS we were thinking it would be good to release a single subcommand called `replace`. but taking on the logic of `replace-many`.  @romain-intel would this CLI interface change vs your PR be palatable?

We could, of course have a `replace` that can do positional arguments and/or the `--add` / `--remove` at the same time.  However that seems a bit ugly and confusing IMO.  My hope is that it's not too late in the game to switch to a `replace-many` type command.

Reply:
It's a breaking API for us internally. I am fine with the replace-many syntax but would definitely prefer an optional thing ie that the replace <to-replace> <replacement> still works. If it is too complicated, we can see about breaking things internally. I do not think this command is very used so it may be ok. I'll check.

Reply:
It's not complicated to code up the args + opts way.  IMO the consideration here is: is this a good UX to expose in OSS (which gets us even deeper into the backward compatibility game, after this point):

* `tag replace --remove R1 --remove R2 --add A1 --add A2  R3 A3`

If we are able to accept the short term annoyance of dropping `R3 A3` above... I think there are long term benefits globally:
* I think it's confusing for CLI users to have to remember the ordering of R2, A2 above.
* Two ways to do the same thing (A1 and A2).

A middle way could be:
* Disallow simultaneous usage of options and positional args (throw error)
* Hide the positional way in usage / help print, so it becomes a grandfathered hidden feature.  New users of this will naturally only use the options.

What is folks' preferences?  cc @romain-intel @savingoyal @tuulos

Reply:
I am okay with the middle way if there are users of this functionality at Netflix.Otherwise we can drop `R3 A3`. @romain-intel can you confirm if there are any users using this feature?